### PR TITLE
handle capabilities secure_boot and nested_virtualization

### LINKED ISSUE #188: Nested virtualization
Add support for the capability NestedVirtualization when converging catlets.

[AB#667](https://dev.azure.com/dbosoft/eae53133-585f-4d55-8e1f-5cba6fd8ce3b/_workitems/edit/667)

### LINKED ISSUE #189: Secure Boot
Add support for the capability SecureBoot when converging catlets.

[AB#668](https://dev.azure.com/dbosoft/eae53133-585f-4d55-8e1f-5cba6fd8ce3b/_workitems/edit/668)

### PR DESCRIPTION
**Pull request by fw2568**
Adds support for converging nested virtualization and fixes issues for secure boot capability to be converged correctly. 
This PR also adds output of capabilities in reverse generated config. 

As capabilities are opt-in either by catlet or parent(s) the capabilities will only be converged if enabled or disabled. 

fixes #188  and fixes #189 

Remarks: The current genepool catlets have a bug for secure boot capabilitity - they use "SecureBoot", but this in not correct, it has to be "secure_boot". Therefore until fixed in the parents, the reverse config logic will always generate a capability for secure_boot even if it the same on parent.

### CODE REVIEW

File: src/core/test/Eryph.VmManagement.Test/ConvergeNestedVirtualizationTests.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using Eryph.VmManagement.Converging;
+using LanguageExt;
+using LanguageExt.Common;
+using Xunit;
+
+namespace Eryph.VmManagement.Test
+{
+    public class ConvergeNestedVirtualizationTests : IClassFixture<ConvergeFixture>
+    {
+        private readonly ConvergeFixture _fixture;
```
[the remark this answers is not in the document]
Comment (fw2568) [reply src/core/test/Eryph.VmManagement.Test/ConvergeNestedVirtualizationTests.cs:12]: good point, fixed

File: src/core/src/Eryph.VmManagement/Converging/ConvergeNestedVirtualization.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        TypedPsObject<VirtualMachineInfo> vmInfo)
+    {
+        var capability = Context.Config.Capabilities?.FirstOrDefault(x => x.Name ==
+            EryphConstants.Capabilities.NestedVirtualization);
+
+        if (capability == null)
+            return vmInfo;
+
+        var onOffState = (capability.Details?.Any(x =>
+            string.Equals(x, "off", StringComparison.InvariantCultureIgnoreCase))).GetValueOrDefault() ? OnOffState.Off : OnOffState.On;
```
[the remark this answers is not in the document]
Comment (fw2568) [reply src/core/src/Eryph.VmManagement/Converging/ConvergeNestedVirtualization.cs:28]: Wow, thanks for finding it! Yes, I completely agree. Changed to enabled, disabled.

File: src/core/src/Eryph.VmManagement/Converging/ConvergeSecureBoot.cs [hunk 3]
```diff
@@ -19,17 +21,18 @@ public ConvergeSecureBoot(ConvergeContext context) : base(context)
...
+            EryphConstants.Capabilities.SecureBoot);
 
         if (secureBootCapability == null)
             return vmInfo;
 
         var templateName =
-            secureBootCapability.Details?.FirstOrDefault(x => x.StartsWith("Template:"))?.Split(':')[1]
-            ?? "Windows";
+            secureBootCapability.Details?.FirstOrDefault(x => x.StartsWith("template:", 
+                StringComparison.CurrentCultureIgnoreCase))?.Split(':')[1]
```
[the remark this answers is not in the document]
Comment (fw2568) [reply src/core/src/Eryph.VmManagement/Converging/ConvergeSecureBoot.cs:31]: typo, but I justed haved noticed that I'm using InvariantCulture in all most all cases wrong: https://stackoverflow.com/questions/492799/difference-between-invariantculture-and-ordinal-string-comparison

File: src/modules/src/Eryph.Modules.ComputeApi/Handlers/GetCatletConfigurationHandler.cs [hunk 4]
```diff
@@ -242,6 +243,64 @@ public GetCatletConfigurationHandler(IStateStore stateStore)
...
+                capabilities.Add(secureBootCap);
+            }
+
+            // reduce capabilities to only those that are not set same on parent
+            foreach (var capabilityConfig in capabilities.ToArray())
+            {
+                var parentCap = parentCaps
+                    .FirstOrDefault(x => x.Name == capabilityConfig.Name);
+                if (parentCap == null) continue;
+                if((parentCap.Details?.Length == 0
```
[the remark this answers is not in the document]
Comment (fw2568) [reply src/modules/src/Eryph.Modules.ComputeApi/Handlers/GetCatletConfigurationHandler.cs]: fixed, also considering now that case may be different between catlet and parent.

**Review by ChristopherMann — APPROVED** [review]

**Pull request merged**